### ISSUE TITLE
Add Check Support for Java 21 Unnamed Variables & Patterns Syntax: IllegalType

### ISSUE THREAD

**Issue by mahfouz72** [checkstyle/checkstyle#15067]
child of #14942 : 

check doc : https://checkstyle.org/checks/coding/illegaltype.html

---

```
PS D:\CS\test> cat config.xml                                                   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="IllegalType">
        </module>
    </module>
</module>
PS D:\CS\test> cat src/Test.java                                                
import java.util.LinkedHashMap;
import java.util.List;

record ColoredPoint(String p, String x, int c) { }
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }

public class Test {
    void test(Object obj) {
        LinkedHashMap _ = new LinkedHashMap<>();     // violation
        if (obj instanceof LinkedHashMap<?,?> _) { }          // violation
    }
}

PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:9:9: Usage of type 'LinkedHashMap' is not allowed. [IllegalType]
[ERROR] D:\CS\test\src\Test.java:10:28: Usage of type 'LinkedHashMap' is not allowed. [IllegalType]
Audit done.
Checkstyle ends with 2 errors.
PS D:\CS\test> 

```

IllegalType Check works as intended when unnamed.

**Comment by rnveach** [comment]
@mahfouz72 Can you show an example using a normal `instanceof` which isn't a pattern. Will it produce a violation as well?

My current thinking is even if we cannot access the unnamed variable, the type is banned and shouldn't be showing up in the code.

**Comment by mahfouz72** [comment]
```
PS D:\CS\test> cat src/Test.java                                      
import java.util.LinkedHashMap;
import java.util.List;

record ColoredPoint(String p, String x, int c) { }
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }

public class Test {
    void test(Object obj) {
        if (obj instanceof LinkedHashMap<?,?>) { }          // should this be violation ?
    }
}
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
Audit done.
PS D:\CS\test> 
```

**Comment by rnveach** [comment]
> show an example using a normal instanceof which isn't a pattern
> if (obj instanceof LinkedHashMap<?,?>) { }
> // should this be violation ?

A lot of my though process is going to follow https://github.com/checkstyle/checkstyle/issues/15010#issuecomment-2181499091 .

This check is more for making types looser, like moving from concrete classes to interfaces like the default classes show. This verbiage in the document seems to resonate with this `Rationale: Helps reduce coupling on concrete classes`.

In this example, without the pattern there is no variable and there is nothing gained by making the type looser. With a parameter, variable, method return, that will eventually be used or assigned somewhere. With the `instanceof` and no pattern it just falls out into space.

`extends` may be a slight gray area. You could say use banning types in `extends` goes against this, but it does directly create the methods in the class. You could also say that the new class being created will be typed somewhere eventually. We are just pre-empting making a list of new classes since we aren't a multi-file check.

IMO, this shouldn't be a violation. However, I don't think it would be a bad idea to create some other check to flag these other areas. I think there is still some spirit left over from this check that it should cause a violation on an `instanceof` with no pattern. The only downside is duplicating the list of classes users will have to do if they want both to be the same in their own config.

> if (obj instanceof LinkedHashMap<?,?> _) { } // violation
> My current thinking is even if we cannot access the unnamed variable, the type is banned and shouldn't be showing up in the code.

I still agree with my earlier statement. Even though unnamed still leads down to my comment above:
> there is nothing gained by making the type looser

It is still basically a variable being created, named or otherwise. We still don't gain anything from making it looser, but since it is a variable with a type, I feel it should still be violated since it was marked as banned.

> IllegalType Check works as intended when unnamed.

I am agreeing. I think we should ensure we have an input, or create one.

**Comment by rnveach** [comment]
I am also adding we need to expand the description of this check at https://checkstyle.org/checks/coding/illegaltype.html#Description . I think it is lacking details for analyzing new syntax, and it is making us work harder on this check.

**Comment by mahfouz72** [comment]
I came back to think more about this and I believe that violations for IllegalTypes with unnamed is a false positive. As a user, I wouldn’t expect a violation in the following example:
```Java
void test(Object obj) {
    LinkedHashMap _ = new LinkedHashMap<>();  // violation
    if (obj instanceof LinkedHashMap<?,?> _) { }  // violation
}
```

As you said, In this case, the `IllegalType` is not being used in a way that contradicts the principle of "program to an interface, not an implementation." Using unnamed variables does not increase coupling to concrete classes. Therefore, it doesn't violate the rationale of the check, which is to help reduce coupling on concrete classes.

> IMO, this shouldn't be a violation.

IMO, if ` if (obj instanceof LinkedHashMap) { } ` shouldn’t be a violation, then `if (obj instanceof LinkedHashMap _) { } `shouldn’t be a violation either. These instances do not represent a usage of the illegal type, and thus should not be considered violations according to the rationale of the check.

> I feel it should still be violated since it was marked as banned.

based on the check description`(Checks that particular classes or interfaces are never used.)` it is marked as banned from "usage" not from appearing any where in the code base. as you suggested we may make a new check that banned the type at all from the code base even if it is not used. but for the current check with the current description I feel that we shouldn't violate the `illegalType` if it can't be used in any way.

If we aren't sure at least I expect to treat the following examples in the same way either both violation or both ok.

```Java
void test(Object obj) {
    if (obj instanceof LinkedHashMap<?,?> _) { }  // this is a  violation
    if (obj instanceof LinkedHashMap<?,?> ) { }  // this is ok
}
```
@rnveach please reconsider this and decide

**Comment by nrmancuso** [comment]
I would also expect these to be treated the same:

```
void test(Object obj) {
    if (obj instanceof LinkedHashMap<?,?> _) { }  // this is a  violation
    if (obj instanceof LinkedHashMap<?,?> ) { }  // this is ok
}
```

We are falling down a rabbit hole here, we should just add a test input and move on.

**Comment by nrmancuso** [comment]
@mahfouz72 please update the description to show an example with the "old" instanceof and instanceof with unnamed variable.

**Comment by nrmancuso** [comment]
I am reconsidering this update...

I am not sure about the "spirit" of this check. Honestly, I feel that since Checkstyle isn't type-aware, this check tries to be "too smart", and it is not clear to me if behavior for `instanceof` is intentional or a miss (false negative).

IMO, this check is better for SonarQube/Intellij Inspections, where we can see if someone is using methods from, say, LinkedHashMap, that are NOT part of the Map interface, and not placing violations on these usages.

After digging through the existing implementation, and rereading the docs, I am leaning towards believing that the `instanceof` behavior is intentional. The list of tokens is a bunch of parameter/variable/annotation field definitions, which shows that we want to check/validate actual "uses" of the type.

If someone just wanted to forbid a type completely, they would use an import check.  I think this check is more concerned about aspects of a given class that could be exposed outside of the class, and actual usages of an IllegalType.

If we aren't calling specific methods on an IllegalType and "using" it in a given class, we should not violate it.

Here is what I am thinking:


```
➜  src cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="IllegalType"/>
    </module>
</module>

➜  src cat Test.java 
import java.util.LinkedHashMap;
import java.util.Map;

public class Test {

    void m(R r, Map m) {

        // "old" syntax
        if (m instanceof LinkedHashMap<?,?>) {
            // violation below
            LinkedHashMap lhm = (LinkedHashMap<?,?>) m;
        }

        // "new" syntax, equivalent to above, I would expect
        // this to be a violation, and it is
        // violation below
        if (m instanceof LinkedHashMap<?,?> lhm) {

        }

        // "old" syntax, with no declaration of variable, not a violation
        if (m instanceof LinkedHashMap<?,?>) {

        }

        // unnamed equivalent to above, should not be a violation, but is a false positive
        if (m instanceof LinkedHashMap<?,?> _) {

        }

        // record pattern, we declare a variable, should be a violation
        if (r instanceof R(LinkedHashMap<?,?> lhm)) {

        }

        // record pattern with unnamed variable, should not be a violation, but is a false postitive
        if (r instanceof R(LinkedHashMap<?,?> _)) {

        }

        // no violation here
        if (r != null && r.getClass().equals(LinkedHashMap.class)) {

        }
    }

}

record R(Map m) {}

➜  src java -jar checkstyle-10.17.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:11:13: Usage of type 'LinkedHashMap' is not allowed. [IllegalType]
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:17:26: Usage of type 'LinkedHashMap' is not allowed. [IllegalType]
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:27:26: Usage of type 'LinkedHashMap' is not allowed. [IllegalType]
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:32:28: Usage of type 'LinkedHashMap' is not allowed. [IllegalType]
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:37:28: Usage of type 'LinkedHashMap' is not allowed. [IllegalType]
Audit done.
Checkstyle ends with 5 errors.

➜  src javac --enable-preview --release 21 Test.java               
Note: Test.java uses preview features of Java SE 21.
Note: Recompile with -Xlint:preview for details.
```

**Comment by romani** [comment]
Import as be star import, so type will not participate in list of imports.

**Comment by nrmancuso** [comment]
> Import as be star import, so type will not participate in list of imports.

Please restate this, I am not sure how this relates to the discussion.

**Comment by nrmancuso** [comment]
We really should not be doing any updates to this check until we clearly define it's purpose, and make sure that we follow the vision of this check.

This issue is blocked by https://github.com/checkstyle/checkstyle/issues/15186, which will not be worked on in the scope of GSOC.
